src/ints/mouse_dos_driver.cpp

```diff
@@ -0,0 +1,1440 @@
+                 static_cast<uint8_t>((address >> 8) & 0xff));
+        IO_Write(cr, 0xf);
+        IO_Write(static_cast<io_port_t>(cr + 1),
+                 static_cast<uint8_t>(address & 0xff));
+    }
+}
+
+// ***************************************************************************
+// Graphic mode cursor
+// ***************************************************************************
+
+static uint8_t gfxReg3CE[9];
+static uint8_t index3C4, gfxReg3C5;
+
+static void SaveVgaRegisters()
+{
+    if (IS_VGA_ARCH) {
+        for (uint8_t i = 0; i < 9; i++) {
+            IO_Write(0x3CE, i);
+            gfxReg3CE[i] = IO_Read(0x3CF);
+        }
+        // Setup some default values in GFX regs that should work
+        IO_Write(0x3CE, 3);
+        IO_Write(0x3Cf, 0); // disable rotate and operation
+        IO_Write(0x3CE, 5);
+        IO_Write(0x3Cf, gfxReg3CE[5] & 0xf0); // Force read/write mode 0
+
+        // Set Map to all planes. Celtic Tales
+        index3C4 = IO_Read(0x3c4);
+        IO_Write(0x3C4, 2);
+        gfxReg3C5 = IO_Read(0x3C5);
+        IO_Write(0x3C5, 0xF);
+    } else if (machine == MCH_EGA) {
+        // Set Map to all planes.
+        IO_Write(0x3C4, 2);
+        IO_Write(0x3C5, 0xF);
+    }
+}
+
+static void RestoreVgaRegisters()
+{
+    if (IS_VGA_ARCH) {
+        for (uint8_t i = 0; i < 9; i++) {
+            IO_Write(0x3CE, i);
+            IO_Write(0x3CF, gfxReg3CE[i]);
+        }
+
+        IO_Write(0x3C4, 2);
+        IO_Write(0x3C5, gfxReg3C5);
+        IO_Write(0x3C4, index3C4);
+    }
+}
+
+static void ClipCursorArea(int16_t &x1, int16_t &x2, int16_t &y1, int16_t &y2,
+                           uint16_t &addx1, uint16_t &addx2, uint16_t &addy)
+{
+    addx1 = 0;
+    addx2 = 0;
+    addy  = 0;
+    // Clip up
+    if (y1 < 0) {
+        addy = static_cast<uint16_t>(addy - y1);
+        y1 = 0;
+    }
+    // Clip down
+    if (y2 > state.clipy) {
+        y2 = state.clipy;
+    };
+    // Clip left
+    if (x1 < 0) {
+        addx1 = static_cast<uint16_t>(addx1 - x1);
+        x1 = 0;
+    };
+    // Clip right
+    if (x2 > state.clipx) {
+        addx2 = static_cast<uint16_t>(x2 - state.clipx);
+        x2    = state.clipx;
+    };
+}
+
+static void RestoreCursorBackground()
+{
+    if (state.hidden || state.inhibit_draw || !state.background)
+        return;
+
+    SaveVgaRegisters();
+
+    // Restore background
+    int16_t x, y;
+    uint16_t addx1, addx2, addy;
+    uint16_t dataPos = 0;
+    int16_t x1       = static_cast<int16_t>(state.backposx);
+    int16_t y1       = static_cast<int16_t>(state.backposy);
+    int16_t x2       = static_cast<int16_t>(x1 + CURSOR_SIZE_X - 1);
+    int16_t y2       = static_cast<int16_t>(y1 + CURSOR_SIZE_Y - 1);
+
+    ClipCursorArea(x1, x2, y1, y2, addx1, addx2, addy);
+
+    dataPos = static_cast<uint16_t>(addy * CURSOR_SIZE_X);
+    for (y = y1; y <= y2; y++) {
+        dataPos = static_cast<uint16_t>(dataPos + addx1);
+        for (x = x1; x <= x2; x++) {
```

Comment:
You could initialise x and y in the loops no?

Reply:
Done.